src/Instrumentation/ExtAmqp/src/ExtAmqpInstrumentation.php

```diff
@@ -0,0 +1,225 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OpenTelemetry\Contrib\Instrumentation\ExtAmqp;
+
+use AMQPExchange;
+use AMQPQueue;
+use OpenTelemetry\API\Globals;
+use OpenTelemetry\API\Instrumentation\CachedInstrumentation;
+use OpenTelemetry\API\Trace\Span;
+use OpenTelemetry\API\Trace\SpanKind;
+use OpenTelemetry\API\Trace\StatusCode;
+use OpenTelemetry\Context\Context;
+use OpenTelemetry\Context\Propagation\ArrayAccessGetterSetter;
+use function OpenTelemetry\Instrumentation\hook;
+use OpenTelemetry\SemConv\TraceAttributes;
+use Throwable;
+
+final class ExtAmqpInstrumentation
+{
+    public const NAME = 'ext_amqp';
+
+    public static function register(): void
+    {
+        $instrumentation = new CachedInstrumentation('io.opentelemetry.contrib.php.ext_amqp');
+
+        hook(
+            AMQPExchange::class,
+            'publish',
+            pre: static function (
+                AMQPExchange $exchange,
+                array $params,
+                string $class,
+                string $function,
+                ?string $filename,
+                ?int $lineno,
+            ) use ($instrumentation): array {
+                $routingKey = $params[1];
+
+                /** @psalm-suppress ArgumentTypeCoercion */
+                $builder = $instrumentation
+                    ->tracer()
+                    ->spanBuilder($routingKey . ' publish')
+                    ->setSpanKind(SpanKind::KIND_PRODUCER)
+                    // code
+                    ->setAttribute(TraceAttributes::CODE_FUNCTION, $function)
+                    ->setAttribute(TraceAttributes::CODE_NAMESPACE, $class)
+                    ->setAttribute(TraceAttributes::CODE_FILEPATH, $filename)
+                    ->setAttribute(TraceAttributes::CODE_LINENO, $lineno)
+                    // messaging
+                    ->setAttribute(TraceAttributes::MESSAGING_SYSTEM, 'amqp')
+                    ->setAttribute(TraceAttributes::MESSAGING_OPERATION, 'publish')
+
+                    ->setAttribute(TraceAttributes::MESSAGING_DESTINATION, $routingKey)
+                    ->setAttribute(TraceAttributes::MESSAGING_DESTINATION_NAME, $routingKey)
+                    ->setAttribute(TraceAttributes::MESSAGING_DESTINATION_PUBLISH_NAME, $routingKey)
```

Comment:
Must use exchange name, [spec](https://github.com/open-telemetry/semantic-conventions/blob/40839cbd53939126b213f2b50921b92a107719e0/docs/messaging/rabbitmq.md#rabbitmq-attributes)
> In RabbitMQ, the destination is defined by an exchange and a routing key. messaging.destination.name MUST be set to the name of the exchange. This will be an empty string if the default exchange is used.

Reply:
Done, thanks